### PR TITLE
Keep headers after an error

### PR DESCRIPTION
**Pull request by PlasmaPower**
If this is merged I'll rewrite it for v2.x.

### CODE REVIEW

**Comment by PlasmaPower** [comment]
Oops, forgot we have to test for node 0.12, where `Object.assign` doesn't exist. Fixing now.

**Comment by fengmk2** [comment]
This feature should be another middleware I think.

**Comment by fengmk2** [comment]
Ok, I see this just want to set cors headers only on error. This should be ok.

And now we should think about performance. Don't bind set function every request. set(this, key, value) will be better. And don't try catch if show header option is false.

**Comment by PlasmaPower** [comment]
Is this good? The set function still has to be local because it needs a local copy of `headersSet`.

File: index.js [hunk 1]
```diff
@@ -49,6 +50,8 @@ module.exports = function (options) {
 
   options.credentials = !!options.credentials;
 
+  options.headersKeptOnError = options.headersKeptOnError === undefined || !!options.headersKeptOnError;
```
Comment (fengmk2) [inline index.js]: should be `keepHeadersOnError`

Comment (PlasmaPower) [inline index.js]: Oops, this was from when I was planning on making it an array so you could only keep some. Will be fixed in just a bit.

Comment (PlasmaPower) [inline index.js]: Fixed.

File: index.js [hunk 2]
```diff
@@ -70,54 +73,76 @@ module.exports = function (options) {
...
       }
-
-      yield next;
     } else {
       // Preflight Request
 
       // If there is no Access-Control-Request-Method header or if parsing failed,
       // do not set any additional headers and terminate this set of steps.
       // The request is outside the scope of this specification.
-      if (!this.get('Access-Control-Request-Method')) {
```
Comment (fengmk2) [inline index.js]: on `Preflight Request`, there is won't throw any error on old code logic https://github.com/koajs/cors/blob/master/index.js#L88.

Comment (PlasmaPower) [inline index.js]: Oh yeah I guess it won't have set any headers. Reverting.

Comment (PlasmaPower) [inline index.js]: Fixed, that should simplify the diff quite a bit.

**Comment by fengmk2** [comment]
+1

**Comment by fengmk2** [comment]
1.2.0 release.

@PlasmaPower please pick this feature into v2.x branch

**Pull request merged**